### PR TITLE
Add external exporter to Pelorus

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Describe the behavior changes introduced in this PR

Add a Proof of Concept (PoC) for adding external exporters to Pelorus.

## Linked Issues

No.

## Testing Instructions

Follow the instructions in `poc_external_exporter/poc_external_prometheus.md`.

### CODE REVIEW

File: poc_external_exporter/prometheus-additional.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+- job_name: "test"
+  static_configs:
+  - targets: ["node.demo.do.prometheus.io"]
```
Comment (mpryc) [inline poc_external_exporter/prometheus-additional.yaml:3]: If there is more then one target it's standard list, so can be?:

```yaml
- job_name: "test"
  static_configs:
  - targets:
    -  "node.demo.do.prometheus.io"
    -  "another.target.io"
```

Comment (mateusoliveira43) [reply poc_external_exporter/prometheus-additional.yaml:3]: I think so https://prometheus.io/docs/prometheus/latest/configuration/configuration/#static_config

**Review by mpryc — CHANGES_REQUESTED** [review]
Overall I would use the already existing helm chart to create secrets and other resources instead of manually creating them:

https://github.com/konveyor/pelorus/blob/master/charts/pelorus/templates/prometheus-scrape-config-secret.yaml

Also some instructions on how this was done:

https://github.com/konveyor/pelorus/commit/f03a03677627349e3ccd38318d942398f78edf10#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R47

**Comment by mateusoliveira43** [comment]
As talked with @mpryc, using `extra_prometheus_hosts` should get the job done. But I tested out and I believe `extra_prometheus_hosts` is not working (the configuration of charts seems correct, `charts/pelorus/templates/prometheus-scrape-config-secret.yaml` would create the config and secret and `charts/pelorus/templates/prometheus-cr.yaml` would add the config to prometheus operator, just like in this PoC, but I think not even the secret is being created, because `additional-scrape-configs` does not appear in `oc get secret -n pelorus`)

I have created an Pelorus instance with
```yaml
apiVersion: charts.pelorus.konveyor.io/v1alpha1
kind: Pelorus
metadata:
  name: external-example-configuration
spec:
  exporters:
    instances:
      - app_name: deploytime-exporter
        exporter_type: deploytime
        extraEnv:
          - name: NAMESPACES
            value: example_namespace
      - app_name: committime-exporter
        exporter_type: committime
        extraEnv:
          - name: NAMESPACES
            value: example_namespace
  extra_prometheus_hosts:
    - id: "mateus"
      hostname: "node.demo.do.prometheus.io"
      password: ""
```
to test it out.

I tried it manually, by creating `prometheus-additional.yaml` with (this is the content that helm would have created using `extra_prometheus_hosts`)
```yaml
- job_name: "federated-prometheus-mateus"
  scrape_interval: 15s
  honor_labels: true
  metrics_path: '/federate'
  params:
    'match[]':
      - '{job="openshift-state-metrics"}'
  scheme: "https"
  basic_auth:
    username: 'internal'
    password: ""
  tls_config:
    insecure_skip_verify: true
  static_configs:
    - targets:
        - "node.demo.do.prometheus.io"
      labels:
        federated_job: "federated-prometheus-mateus"
```
and following the subsequent steps of `poc_external_exporter/poc_external_prometheus.md`, and it works.
Screenshot from 2023-03-04 12-23-29

I suspect is something related to checking if the variable is being used, the other logic seems correct.

**Comment by mateusoliveira43** [comment]
Tested more, and I think the problem is the operator, not helm.

I have run `helm template --debug pelorus-test charts/pelorus > debug.yaml` changing the `charts/pelorus/values.yaml`, and the configs are added to `prometheus-cr.yaml`.

I tried to used `custom_ca` in operator, and it did not change anything in `prometheus-cr.yaml` (to test this, create an instance, can be like the example in OpenShift, but adding `custom_ca: true`. It should apply the logic of `charts/pelorus/templates/prometheus-cr.yaml`, but if you check the object, it does not have any of the options `custom_ca` would add).

I believe the problem is in `pelorus-operator/bundle/manifests/charts.pelorus.konveyor.io_pelorus.yaml` (related to https://github.com/konveyor/pelorus/issues/751). Because the values of `spec.properties` that are listed (only `exporters`, `prometheus_retention`, `prometheus_retention_size`, `prometheus_storage`, `prometheus_storage_pvc_capacity` and  `prometheus_storage_pvc_storageclass` from [full list](https://pelorus.readthedocs.io/en/v2.0.7/GettingStarted/configuration/PelorusCore/#list-of-all-configuration-options)), seem to work, but `extra_prometheus_hosts` and `custom_ca`, for example, that are not listed, do not work.

**Comment by mpryc** [comment]
@mateusoliveira43 tested and you are right the values are not consumed by the operator.

The Prometheus operator expected values [additionalScrapeConfigs](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/api.md#monitoring.coreos.com/v1.PrometheusSpec) which are in our templates are [correct](https://github.com/konveyor/pelorus/blob/master/charts/pelorus/templates/prometheus-cr.yaml#L39)

The pelorus operator templates fails to create secret `additional-scrape-configs` from the `charts/pelorus/templates/prometheus-scrape-config-secret.yaml`, which then is used by the prometheus-operator.

**Comment by mpryc** [comment]
@mateusoliveira43 looks like it's not only related to the external exporter configs, also defining things such as thanos won't be used:

```
thanos_bucket_name: <...>
bucket_access_point: <...>
bucket_access_key: <...>
bucket_secret_access_key: <...>
```

This did not created proper thanos configs and configured grafana to point to the prometheus rather then thanos.

**Comment by mpryc** [comment]
The top level configs such as, does work:

```
  prometheus_retention: 1y
  prometheus_retention_size: 1GB
  prometheus_storage: true
  prometheus_storage_pvc_capacity: 2Gi
  prometheus_storage_pvc_storageclass: gp2
```

**Comment by mpryc** [comment]
So it looks like anything outside of the schama is ignored while creating instance. This was not the case when we created 0.0.1 of this operator.

**Comment by mpryc** [comment]
Depends On: https://github.com/konveyor/pelorus/pull/861

Note it's possible to test without creating new operator by modyfing `Pelorus` CRD in the cluster.
Left side menu (Administration->CustomResourceDefinition), search for Pelorus, select Yaml and add missing option (number of spaces are important).

Screenshot from 2023-03-06 13-07-10

**Comment by mateusoliveira43** [comment]
@mpryc this workaround did not work for me, the Pelorus manager gets error

**Comment by mateusoliveira43** [comment]
Closing this PR without merging as it was a PoC. The implementation with the given requirements from https://github.com/orgs/konveyor/projects/51/views/2?pane=issue&itemId=20613515 will be opened in follow up

**Pull request closed without merging**